uPortal-security/uPortal-security-core/src/main/java/org/apereo/portal/security/oauth/IdTokenFactory.java

```diff
@@ -0,0 +1,265 @@
+
+    /** JSON data type 'string' */
+    @Value("${org.apereo.portal.security.oauth.IdTokenFactory.mapping.zoneinfo:}")
+    private String zoneinfoAttr;
+
+    /** JSON data type 'string' */
+    @Value("${org.apereo.portal.security.oauth.IdTokenFactory.mapping.locale:}")
+    private String localeAttr;
+
+    /** JSON data type 'string' */
+    @Value(
+            "${org.apereo.portal.security.oauth.IdTokenFactory.mapping.phone_number:telephoneNumber}")
+    private String phoneNumberAttr;
+
+    /** JSON data type 'boolean' */
+    @Value("${org.apereo.portal.security.oauth.IdTokenFactory.mapping.phone_number_verified:}")
+    private String phoneNumberVerifiedAttr;
+
+    /*
+     * NB:  The 'address' claim requires additional complexity b/c it's type is JSON object.
+     */
+
+    /** JSON data type 'number' */
+    @Value("${org.apereo.portal.security.oauth.IdTokenFactory.mapping.updated_at:}")
+    private String updatedAtAttributeName;
+
+    private List<ClaimMapping> mappings;
+
+    @PostConstruct
+    public void init() {
+        final List<ClaimMapping> list = new ArrayList<>();
+
+        list.add(new ClaimMapping("sub", subAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("name", nameAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("given_name", givenNameAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("family_name", familyNameAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("middle_name", middleNameAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("nickname", nicknameAttr, DataTypeConverter.STRING));
+        list.add(
+                new ClaimMapping(
+                        "preferred_username", preferredUsernameAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("profile", profileAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("picture", pictureAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("website", websiteAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("email", emailAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("email_verified", emailVerifiedAttr, DataTypeConverter.BOOLEAN));
+        list.add(new ClaimMapping("gender", genderAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("birthdate", birthdateAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("zoneinfo", zoneinfoAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("locale", localeAttr, DataTypeConverter.STRING));
+        list.add(new ClaimMapping("phone_number", phoneNumberAttr, DataTypeConverter.STRING));
+        list.add(
+                new ClaimMapping(
+                        "phone_number_verified",
+                        phoneNumberVerifiedAttr,
+                        DataTypeConverter.BOOLEAN));
+        list.add(new ClaimMapping("updated_at", updatedAtAttributeName, DataTypeConverter.NUMBER));
+
```

Comment:
Does anyone know if it's permitted and/or sane to add claims to the OIDC ID Token that link the user to roles or groups?  @jtgasper3?

Reply:
Per the discussion we had, yes it is permitted (something like `“roles”: ["Admin", … ]`), but I'd recommend some value filtering be in-place to limit values to application associated groups/roles, otherwise token bloat can be an issue. MemberOf can have hundreds of values and will often include the full DN in the value, if being pulled directly from LDAP, which probably provides very little value.